test/e2e/auth/node_authn.go

```diff
@@ -0,0 +1,127 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/master/ports"
+	"k8s.io/kubernetes/test/e2e/framework"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = SIGDescribe("[Feature:NodeAuthenticator]", func() {
+
+	f := framework.NewDefaultFramework("node-authn")
+	var ns string
+	var nodeIPs []string
+	BeforeEach(func() {
+		ns = f.Namespace.Name
+
+		nodeList, err := f.ClientSet.CoreV1().Nodes().List(metav1.ListOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(len(nodeList.Items)).NotTo(BeZero())
+
+		pickedNode := nodeList.Items[0]
+		nodeIPs = framework.GetNodeAddresses(&pickedNode, v1.NodeExternalIP)
+		nodeIPs = append(nodeIPs, framework.GetNodeAddresses(&pickedNode, v1.NodeInternalIP)...)
+
+		// make sure ServiceAccount admission controller is enabled, so secret generation on SA creation works
+		sa, err := f.ClientSet.CoreV1().ServiceAccounts(ns).Get("default", metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(len(sa.Secrets)).NotTo(BeZero())
+	})
+
+	It("The kubelet's main port 10250 should fail with no credentials", func() {
+		pod := createNodeAuthTestPod(f, ns)
+		for _, nodeIP := range nodeIPs {
+			// Anonymous authentication is disabled by default
+			_, err := framework.LookForStringInPodExec(pod.Namespace,
+				pod.Name,
+				[]string{"curl",
+					"--insecure",
+					fmt.Sprintf("https://%s:%v/metrics", nodeIP, ports.KubeletPort),
+				},
+				"Unauthorized",
+				time.Minute)
+			Expect(err).NotTo(HaveOccurred())
+		}
+	})
+
+	It("The kubelet can delegate ServiceAccount tokens to the API server", func() {
+		By("create a new ServiceAccount for authentication")
+		falseValue := false
+		newSA := &v1.ServiceAccount{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: ns,
+				Name:      "node-auth-newSA",
+			},
+			AutomountServiceAccountToken: &falseValue,
+		}
+		sa, err := f.ClientSet.CoreV1().ServiceAccounts(ns).Create(newSA)
+		Expect(err).NotTo(HaveOccurred())
+		saSecret, err := f.ClientSet.CoreV1().Secrets(ns).Get(sa.Secrets[0].Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		pod := createNodeAuthTestPod(f, ns)
+
+		for _, nodeIP := range nodeIPs {
+			_, err := framework.LookForStringInPodExec(pod.Namespace,
+				pod.Name,
+				[]string{"curl",
+					"--insecure",
+					fmt.Sprintf(`--header "Authorization: Bearer %s"`, string(saSecret.Data["token"])),
+					fmt.Sprintf("https://%s:%v/metrics", nodeIP, ports.KubeletPort),
+				},
+				"Forbidden",
```

Comment:
I get `Unauthorized`. I think a more robust approach would be to just look for a 401 or 403 status code:

```
curl -s -o /dev/null -w "%{http_code}" ... | grep "^40[13]$"
```

Then you can just check the exit status of exec command too.